src/main/java/com/salesforce/functions/jvm/sdk/data/bulk/BulkApi.java

```diff
@@ -0,0 +1,276 @@
+   *     'sf__Id' fields.
+   * @throws DataApiException If an API error occurred during the operation.
+   * @see <a
+   *     href="https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/get_job_failed_results.htm">API
+   *     Reference</a>
+   * @see BulkApi#getSuccessfulRecordResults(IngestJobReference)
+   * @see BulkApi#getUnprocessedRecordResults(IngestJobReference)
+   */
+  @Nonnull
+  @SuppressWarnings("unused")
+  DataTable getFailedRecordResults(IngestJobReference reference) throws DataApiException;
+
+  /**
+   * Retrieves a list of unprocessed records for failed or aborted ingest jobs.
+   *
+   * @param reference The reference of the job to retrieve the list for.
+   * @return A {@link DataTable} with the original uploaded data.
+   * @throws DataApiException If an API error occurred during the operation.
+   * @see <a
+   *     href="https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/get_job_unprocessed_results.htm">API
+   *     Reference</a>
+   * @see BulkApi#getSuccessfulRecordResults(IngestJobReference)
+   * @see BulkApi#getFailedRecordResults(IngestJobReference)
+   */
+  @Nonnull
+  @SuppressWarnings("unused")
+  DataTable getUnprocessedRecordResults(IngestJobReference reference) throws DataApiException;
+
+  /**
+   * Gets the results for a query job. The job must be in the complete state.
+   *
+   * @param reference The reference of the job to get the results for.
+   * @param maxRecords The maximum number of records to retrieve.
+   * @return A {@link QueryJobResult} that can be used to get the result data as well as to fetch
+   *     the next page or results.
+   * @throws DataApiException If an API error occurred during the operation.
+   * @see BulkApi#getMoreQueryResults(QueryJobResult)
+   * @see <a
+   *     href="https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/query_get_job_results.htm">API
+   *     Reference</a>
+   */
+  @Nonnull
+  @SuppressWarnings("unused")
+  QueryJobResult getQueryJobResults(QueryJobReference reference, long maxRecords)
+      throws DataApiException;
+
+  /**
+   * Gets the next results for a query job.
+   *
+   * @param result The previous query job result.
+   * @return A {@link QueryJobResult} that can be used to get the result data as well as to fetch
+   *     the next page or results.
+   * @throws DataApiException If an API error occurred during the operation.
+   * @see BulkApi#getQueryJobResults(QueryJobReference, long)
+   * @see <a
+   *     href="https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/query_get_job_results.htm">API
+   *     Reference</a>
+   */
+  @Nonnull
+  @SuppressWarnings("unused")
+  QueryJobResult getMoreQueryResults(QueryJobResult result) throws DataApiException;
+
+  @SuppressWarnings("unused")
+  void waitForJobState(QueryJobReference reference, IngestJobState state)
```

Comment:
i don't think it's useful to wait for an explicit state.  you don't need to monitor or wait for any state until the job is in the`UploadComplete` state.  the state transitions from there are always:
* `Aborted`
* `Failed`
* `InProgress` and this state can only transition into:
  * `Aborted`
  * `Failed`
  * `JobComplete`

so the user almost every time will want to wait for a state `Aborted`, `Failed`, and `JobComplete`.  the failure states are also valid for retrieving job information.  a state of `Aborted` or `Failed` should not throw any errors.  

this current method signature cannot achieve this.  i say keep it just as `waitForJobState(QueryJobReference reference)`

Reply:
correction.  if dropping the state argument the method should be named `waitForJobCompletion`